Multiprotocol/FrSkyR9_sx1276.ino

```diff
@@ -0,0 +1,238 @@
+
+	uint8_t val = SX1276_ReadReg(REG_DETECT_OPTIMIZE);
+	val = (val & 0b11111000) | 0b00000101;
+	SX1276_WriteReg(REG_DETECT_OPTIMIZE, val);
+
+	// val = SX1276_ReadReg(REG_MODEM_CONFIG2);
+	// val = (val & 0b00011111) | 0b11000000;
+	// writeRegister(REG_MODEM_CONFIG2, val);
+
+	SX1276_WriteReg(REG_MODEM_CONFIG1, 0x93);
+
+	SX1276_WriteReg(REG_MODEM_CONFIG2, 0x60);
+
+	val = SX1276_ReadReg(REG_MODEM_CONFIG3);
+	val = (val & 0b11110011);
+	SX1276_WriteReg(REG_MODEM_CONFIG3, val);
+
+	SX1276_WriteReg(REG_PREAMBLE_LSB, 9);
+
+	SX1276_WriteReg(REG_DETECTION_THRESHOLD, 0x0C);
+
+	SX1276_WriteReg(REG_LNA, 0x23);
+
+	SX1276_WriteReg(REG_HOP_PERIOD, 0x00);
+
+	val = SX1276_ReadReg(REG_PA_DAC);
+	val = (val & 0b11111000) | 0b00000111;
+	SX1276_WriteReg(REG_PA_DAC, val);
+
+	// TODO this can probably be shorter
+	return 20000; // start calling FrSkyR9_callback in 20 milliseconds
+}
+
+uint16_t FrSkyR9_callback()
+{
+    static uint16_t index = 0;
+	uint8_t buffer[3];
+
+	SX1276_WriteReg(REG_OP_MODE, 0x81); // STDBY
+	SX1276_WriteReg(REG_IRQ_FLAGS_MASK, 0xbf); // use only RxDone interrupt
+
+	buffer[0] = 0x00;
+	buffer[1] = 0x00;
+	SX1276_WriteRegisterMulti(REG_DIO_MAPPING1, buffer, 2); // RxDone interrupt mapped to DIO0 (the rest are not used because of the REG_IRQ_FLAGS_MASK)
+
+	// SX1276_WriteReg(REG_PAYLOAD_LENGTH, 13);
+
+	// SX1276_WriteReg(REG_FIFO_ADDR_PTR, 0x00);
+
+	// SX1276_WriteReg(REG_OP_MODE, 0x85); // RXCONTINUOUS
+	// delay(10); // 10 ms
+
+	// SX1276_WriteReg(REG_OP_MODE, 0x81); // STDBY
+
+	SX1276_WriteReg(REG_PA_CONFIG, 0xF0);
+
+	uint32_t freq = _freq_map[index] / 61;
+	
+	buffer[0] = (freq & (0xFF << 16)) >> 16;
+	buffer[1] = (freq & (0xFF << 8)) >> 8;
+	buffer[2] = freq & 0xFF;
+	
+	SX1276_WriteRegisterMulti(REG_FRF_MSB, buffer, 3); // set current center frequency
+	
+	delayMicroseconds(500);
+
+	SX1276_WriteReg(REG_PAYLOAD_LENGTH, 26);
+	SX1276_WriteReg(REG_FIFO_TX_BASE_ADDR, 0x00);
+	SX1276_WriteReg(REG_FIFO_ADDR_PTR, 0x00);
+
+	uint8_t payload[26];
+
+	payload[0] = 0x3C; // ????
+	payload[1] = rx_tx_addr[3]; // unique radio id
+	payload[2] = rx_tx_addr[2]; // unique radio id
+	payload[3] = index; // current channel index
+	payload[4] = _step; // step size and last 2 channels start index
+	payload[5] = RX_num; // receiver number from OpenTX
+
+	// binding mode: 0x00 regular / 0x41 bind?
+	if(IS_BIND_IN_PROGRESS)
+		payload[6] = 0x41; 
+	else
+		payload[6] = 0x00;
+
+	// TODO
+	payload[7] = 0x00; // fail safe related (looks like the same sequence of numbers as FrskyX protocol)
+
+	// two channel are spread over 3 bytes.
+	// each channel is 11 bit + 1 bit (msb) that states whether
+	// it's part of the upper channels (9-16) or lower (1-8) (0 - lower 1 - upper)
+
+	const int payload_offset = 8;
+	const bool is_upper = false;
+
+	int chan_index = 0;
+
+	for(int i = 0; i < 8; i += 3)
+	{
+		// map channel values (0-2047) to (64-1984)
+		uint16_t ch1 = 64 + (uint16_t)((1920.0f / 2047.0f) * Channel_data[chan_index]);
```

Comment:
Can you use the existing converstion function instead of floating point calculation: FrSkyX_scaleForPXX()

Reply:
It does the same thing? Should I use it as is or use `convert_channel_frsky` like in the the FrSkyX protocol?